README.txt
Various tools used for backups
==============================

Currently contains two utilities:

  1. tarbackup.py: Essentially a wrapper for gnu tar. It uses the incremental
     backup options, and makes a full backup every N days, where you can define
     how long such a cycle should be. Each backup goes in a uniquely named
     tarball for easier management. Never overwrites previous files.

  2. ftpsync.py: A tool that synchronises a local directory with one on an
     ftp server. Uses a very simple protocol of only checking file names
     and file sizes. File sizes are not allowed to change. Combined with
     tarbackup.py, this will essentially upload your tarballs to an ftp server,
     adding any new incremental or full backups that were created locally,
     and deleting any that were removed locally. FTP login details uses a
     conventional .netrc file in your home directory.

Use with zope/repozo
--------------------

ftpsync.py has been optimised to work with the zodb backup tool named repozo.
Because repozo also leaves old files unmodified, and only creates an
incremental file for whatever has changed since the previous backup, this fits
right in with our policy. The only exception is that repozo does update a .dat
file. To accomodate this, ftpsync.py has a --always option whereby you can
instruct it to always resync .dat files.

Example .netrc file
--------------------
machine ftpbackup.yourserver.tld login username password secret


Example shell script
--------------------

#!/bin/sh -e
SERVER="ftpbackup.yourserver.tld"

# /etc
/usr/bin/python /usr/bin/tarbackup.py --prefix=myserver_etc \
    --snapshot=/backups/snapshot --target=/backups/other /etc

# zope
sudo -u zope /path/to/zope/bin/repozo -BvzQ -f \
    "/path/to/zope/var/filestorage/Data.fs" -r /backups/zope

# Postgresql
stamp=`date +%s`
weekday=`date +%w`
sudo -u postgres pg_dumpall | gzip > /backups/pgsql/$weekday.$stamp.sql.gz
find "/backups/pgsql" -type f -name "$weekday.*.sql.gz" \
  -a -not -name "$weekday.$stamp.sql.gz" | xargs -r rm -f
  
/usr/bin/python /usr/bin/ftpsync.py --server $SERVER \
    --local /backups/other --remote other
/usr/bin/python /usr/bin/ftpsync.py --server $SERVER \
    --local /backups/zope --remote zope --always="*.dat"
/usr/bin/python /usr/bin/ftpsync.py --server $SERVER \
    --local /backups/pgsql --remote pgsql

# Notify nagios (passive check). This step will only be reached if nothing
# fails above, because sh is called with -e.
echo -e "myhost\tbacktup\tOK:completed" | \
  /usr/sbin/send_nsca -H monitor.host.tld -c /etc/send_nsca.cfg


debian/changelog
backuptools (6) unstable; urgency=low

  * Add AES decryption tool.

 -- Izak Burger <izak@upfrontsystems.co.za>  Mon, 01 Jul 2013 12:44:23 +0200

backuptools (5) unstable; urgency=low

  * Add AES encryption feature.

 -- Izak Burger <izak@upfrontsystems.co.za>  Thu, 26 Jun 2013 12:11:21 +0200

backuptools (4) unstable; urgency=low

  * Show filename that caused assertion error for size check.

 -- Izak Burger <izak@upfrontsystems.co.za>  Thu, 08 Oct 2009 09:39:26 +0200

backuptools (3) unstable; urgency=low

  * makebackupscript.py doesn't crash if it fails to parse a zope.conf file

 -- Izak Burger <izak@upfrontsystems.co.za>  Tue, 25 Aug 2009 15:06:38 +0200

backuptools (2) unstable; urgency=low

  * makebackupscript.py uses system zope install

 -- Izak Burger <izak@upfrontsystems.co.za>  Mon, 20 Jul 2009 16:58:08 +0200

backuptools (1) unstable; urgency=low

  * Initial release

 -- Izak Burger <izak@upfrontsystems.co.za>  Fri, 10 Jul 2009 12:26:56 +0200


decrypt.py
#!/usr/bin/env python

import sys
import os
import struct
import logging
from hashlib import sha256
from optparse import OptionParser

try:
    from Crypto.Cipher import AES
    HAS_CRYPTO=True
except ImportError:
    HAS_CRYPTO=False

class DecryptedFile(object):
    """ Wrapper object that takes a file object, decrypts it using AES. File
        should be prepended with the text bkptools, followed by 8 bytes for the
        file size. """
    def __init__(self, key, iv, fp):
        self.crypter = AES.new(key, AES.MODE_CBC, iv)
        self.fp = fp

        # Read the file header, check that this is one of ours and retrieve
        # the file size
        head = struct.unpack('<QQ', fp.read(16))
        assert struct.pack('<Q', head[0]) == 'bkptools', \
            "Not a backuptools file"
        self.filesize = head[1]

    def read(self):
        # blocksize needs to be a multiple of 16.
        blocksize = 8192
        block = self.fp.read(blocksize)

        # Check for EOF
        if not block:
            return ''

        decrypted = self.crypter.decrypt(block)
        self.filesize -= len(decrypted)

        # Check for last padded block. If self.filesize becomes less than
        # zero, it means the input was not a multiple of 16 and we padded some
        # zeroes onto the file. Strip those off now.
        if self.filesize < 0:
            return decrypted[:self.filesize]

        return decrypted

    def close(self):
        self.fp.close()

def main(args):
    # option parsing
    usage = "%prog --key keyfile filename"
    parser = OptionParser(usage=usage)
    parser.add_option("-k", "--key",
        help="key file for AES decryption. "
             "The last 16 characters of the file name is used as the IV, "
             "padded with nulls if the filename is less than 16 characters")
    parser.add_option("-v", "--verbose", action="count",
        help="Increase verbosity", default=0)

    (options, args) = parser.parse_args()
    if options.key is None:
        parser.print_help(file=sys.stderr)
        sys.exit(1)

    if options.key and not HAS_CRYPTO:
        print >>sys.stderr, "pycrypto not available"
        sys.exit(1)

    # Configure logging
    logging.basicConfig(level=max(4-options.verbose,1)*10)

    # Decrypt the file and spit it out to stdout
    f = args[0]
    key = open(options.key, 'r').read()
    iv = os.path.basename(f)[-16:].ljust(16, '\0')
    fp = DecryptedFile(sha256(key).digest(), iv, open(f, "r"))
    block = fp.read()
    while block:
        sys.stdout.write(block)
        block = fp.read()
    fp.close()

if __name__ == '__main__':
    main(sys.argv)


tarbackup.py
#!/usr/bin/env python

import sys
import os
import logging
import time
from optparse import OptionParser
from glob import glob

DEFAULTCYCLE = 7
DEFAULTPREFIX = "backup"

class BackupFailure(Exception):
    pass

def tarbackup(snapshot, destination, dirs, excludes):
    args = ["tar", "-z", "-c", "--listed-incremental", snapshot, "-f", destination]
    for exclude in excludes:
        args.append("--exclude=%s" % exclude)
    status = os.spawnvp(os.P_WAIT, args[0], args + dirs)
    if status == 2:
        # tar returns 2 for a critical failure
        raise BackupFailure("failed for arguments %s", " ".join(args))

def cleanup(path, pattern, exclude):
    includepattern = os.path.join(path, pattern)
    excludepattern = os.path.join(path, exclude)
    excluded_files = glob(excludepattern)
    files = [f for f in glob(includepattern) if f not in excluded_files]
    for f in files:
        logging.debug("Deleting %s" % f)
        os.remove(f)

def main():
    # option parsing
    usage = "%prog --prefix mybackupname --target /var/backup --snapshot dir dir1 [ dir2 ... ]"
    parser = OptionParser(usage=usage)
    parser.add_option("-p", "--prefix", help="A unique prefix for this backup",
                      default=DEFAULTPREFIX)
    parser.add_option("-t", "--target", help="Where to store the tarball files")
    parser.add_option("-s", "--snapshot", help="Where to store the snapshot files")
    parser.add_option("-c", "--cycle", help="Make a full backup this often (in days)",
                      default=DEFAULTCYCLE, type="int")
    parser.add_option("-e", "--exclude", help="Files to exclude from the backup",
                      default=[], action="append")
    parser.add_option("-v", "--verbose", action="count", help="Increase verbosity",
                      default=0)

    (options, args) = parser.parse_args()

    # Check all required parameters
    for option in (options.target, options.snapshot):
        if option is None:
            parser.print_help()
            sys.exit(1)

    # We need at least one thing to back up
    if len(args) < 1:
        parser.print_help()
        sys.exit(1)

    # Configure logging
    logging.basicConfig(level=max(4-options.verbose,1)*10)

    # Calculate filename and current cycle
    timestamp = int(time.time())
    rightnow = time.localtime(timestamp)
    currentcycle = rightnow[7] / options.cycle

    snapshotfile = "snapshot.%s.%03d" % (options.prefix, currentcycle)
    snapshotpath = os.path.join(options.snapshot, snapshotfile)

    targetfile = "%s.%03d.%d.tar.gz" % (options.prefix, currentcycle, timestamp)
    targetpath = os.path.join(options.target, targetfile)

    # Create paths if they do not exist
    for p in (options.snapshot, options.target):
        if not os.path.exists(p):
            os.makedirs(p)

    # Call tar to do the deed
    logging.info("Running tar backup for %s" % ", ".join(args))
    logging.info("snapshot at %s, destination at %s" % (snapshotpath, targetpath))
    tarbackup(snapshotpath, targetpath, args, options.exclude)

    # Cleanup
    logging.info("Cleaning up")
    cleanup(options.snapshot, "snapshot.%s.*" % options.prefix, snapshotfile)
    cleanup(options.target, "%s.*" % options.prefix,
        "%s.%03d.*.tar.gz" % (options.prefix, currentcycle))

if __name__ == '__main__':
    main()


ftpsync.py
#!/usr/bin/env python

import sys
import os
import struct
import logging
from hashlib import sha256
from ftplib import FTP, error_perm
from netrc import netrc
from optparse import OptionParser
from fnmatch import fnmatch

try:
    from Crypto.Cipher import AES
    HAS_CRYPTO=True
except ImportError:
    HAS_CRYPTO=False

class EncryptedFile(object):
    """ Wrapper object that takes a file object, encrypts it using AES. File
        is prepended with th e text bkptools, followed by 8 bytes for the file
        size. To get back the file size, decrypt the first 16 bytes of the file
        using: struct.unpack('<QQ', bytes) and use the second number of the
        tuple. """
    def __init__(self, key, iv, fp):
        self.crypter = AES.new(key, AES.MODE_CBC, iv)
        self.fp = fp

        fp.seek(0, os.SEEK_END)
        self.filesize = fp.tell()
        fp.seek(0)

        self.firstblock = True

    def read(self, blocksize=-1):
        # blocksize needs to be a multiple of 16. The default passed by
        # ftplib (8192) is a multiple of 16. If you override it, make sure your
        # block is also a multiple of 16. Otherwise this will break.

        # The first block of the file is 16 bytes indicating the size
        if self.firstblock:
            self.firstblock = False
            return 'bkptools' + struct.pack('<Q', self.filesize)

        block = self.fp.read(blocksize)
        if not block:
            return ''

        # Pad block up to 16 bytes if necessary
        l = len(block) & 15
        if l:
            padding = (16-l) * '\0'
            block += padding
            
        return self.crypter.encrypt(block)

    def close(self):
        self.fp.close()

def ftp_login(host):
    login, account, password = netrc().authenticators(host)
    server = FTP(host)
    server.login(login, password, account)
    return server

def cd_or_mkdir(server, subdir=None):
    if subdir:
        try:
            server.cwd(subdir)
        except error_perm:
            server.mkd(subdir)
            server.cwd(subdir)


def list_remote_files(server):
    files = server.nlst('.')
    files = [x for x in files if x not in ('.', '..')]
    files.sort()
    return files

def list_local_files(dir):
    files = os.listdir(dir)
    files.sort()
    return files


def check_backup(server, localpath, files, encryption=False):
    # Make sure we are in binary mode
    server.voidcmd("TYPE I")
    for f in files:
        status = os.stat(os.path.join(localpath, f))
        localsize = status.st_size
        remotesize = server.size(f)

        # If encryption is in use, the remote file must be larger than the
        # local one at the very least
        if encryption:
            assert localsize <= remotesize, f
            continue

        # This will cause the process to quit if the sizes mismatch, and
        # the return status will be 1.
        assert localsize == remotesize, f

def main(args):
    # option parsing
    usage = "%prog --server ftp.host.com --local /var/backup --remote backup"
    parser = OptionParser(usage=usage)
    parser.add_option("-s", "--server",
        help="hostname:[port], hostname and optional port for ftp server")
    parser.add_option("-l", "--local",
        help="local path to backup")
    parser.add_option("-r", "--remote",
        help="remote path to backup to, without leading /")
    parser.add_option("-a", "--always",
        help="files that should always be copied/overwritten")
    parser.add_option("-k", "--key",
        help="key file for AES encryption, if you want to use encryption. "
             "The last 16 characters of the file name is used as the IV, "
             "padded with nulls if the filename is less than 16 characters")
    parser.add_option("-v", "--verbose", action="count",
        help="Increase verbosity", default=0)

    (options, args) = parser.parse_args()
    for option in (options.server, options.local):
        if option is None:
            parser.print_help()
            sys.exit(1)

    if options.key and not HAS_CRYPTO:
        print >>sys.stderr, "pycrypto not available"
        sys.exit(1)

    # Configure logging
    logging.basicConfig(level=max(4-options.verbose,1)*10)

    logging.info("Logging on to %s", options.server)
    server = ftp_login(options.server)
    logging.info("Changing/creating %s" % options.remote)
    cd_or_mkdir(server, options.remote)

    # Get a list of remote files
    logging.info("Obtaining list of files in %s" % (options.remote or "."))
    remote_files = list_remote_files(server)

    # Get a list of local files
    logging.info("Obtaining list of files in %s" % options.local)
    tmplist = list_local_files(options.local)

    # Filter out directories
    local_files = []
    for f in tmplist:
        if os.path.isdir(os.path.join(options.local, f)):
            logging.warn("Cannot handle directory %s" % f)
        else:
            local_files.append(f)

    # Calculate what to upload and delete
    logging.info("Calculating difference")
    to_upload = [f for f in local_files if \
                    f not in remote_files \
                    or options.always and fnmatch(f, options.always)]
    to_delete = [f for f in remote_files if f not in local_files]
    logging.info("%d new files on local end" % len(to_upload))
    logging.info("%d files to delete on remote end" % len(to_delete))

    # Delete files removed locally
    for f in to_delete:
        logging.debug("Deleting %s" % f)
        server.delete(f)

    # Copy new files
    for f in to_upload:
        logging.debug("Storing %s" % f)
        fp = open(os.path.join(options.local, f), "r")
        if options.key:
            key = open(options.key, 'r').read()
            iv = f[-16:].ljust(16, '\0')
            fp = EncryptedFile(sha256(key).digest(), iv, fp)
        server.storbinary('STOR %s' % f, fp)
        fp.close()

    # Now check the backup
    logging.info("Checking the backup")
    check_backup(server, options.local, local_files,
        encryption=(options.key is not None))

    server.quit()

if __name__ == '__main__':
    main(sys.argv)
